File: MVideo/ProductsStorage.py
from typing import Generator, Any
from models.InputLine import InputLine


class ProductsStorage:
    __PRODUCTS_COUNT = 1_200_100
    __products: list[str] = list()
    __products_int_ids: dict[str, int] = dict()
    products_recommendations: dict[int, dict[float, list[int]]] = dict()

    def __init__(self, input_lines: Generator[InputLine, Any, None]):
        for line in input_lines:
            self.__add_line_to_storage(line)

    def get_product_str_id(self, int_id: int):
        return self.__products[int_id]

    def get_product_int_id(self, str_id: str):
        if str_id in self.__products_int_ids:
            return self.__products_int_ids[str_id]

        return -1

    def __add_line_to_storage(self, input_line: InputLine):
        product_int_id = self.__init_product(input_line.product_id)
        recommendation_int_id = self.__init_product(input_line.recommendation_id)
        if product_int_id not in self.products_recommendations:
            self.products_recommendations[product_int_id] = dict()
        if input_line.rank not in self.products_recommendations[product_int_id]:
            self.products_recommendations[product_int_id][input_line.rank] = list()
        self.products_recommendations[product_int_id][input_line.rank].append(recommendation_int_id)

    def __init_product(self, product_id: str):
        if product_id in self.__products_int_ids:
            return self.__products_int_ids[product_id]
        product_index = len(self.__products)
        self.__products_int_ids[product_id] = product_index
        self.__products.append(product_id)
        return product_index

File: MVideo/c#/TestMVideo/ProductsStorage.cs
#region using

using System.Collections.Generic;
using TestMVideo.Models;

#endregion

namespace TestMVideo
{
    public class ProductsStorage
    {
        private const int ProductsCount = 1_200_000;
        private readonly List<string> products = new(ProductsCount);
        private readonly Dictionary<string, int> productsStr = new(ProductsCount);

        public ProductsStorage(IEnumerable<InputLine> inputLines)
        {
            foreach (var inputLine in inputLines)
                AddLineToStorage(inputLine);
        }

        public Dictionary<int, Dictionary<float, List<int>>> ProductsWithRecommendations { get; } = new(ProductsCount);

        public string GetProductsStringId(int intId)
        {
            return products[intId];
        }

        public int GetProductsIntId(string stringId)
        {
            return productsStr.GetValueOrDefault(stringId, -1);
        }

        private void AddLineToStorage(InputLine inputLine)
        {
            var productIntId = InitProduct(inputLine.ProductId);
            var recommendationIntId = InitProduct(inputLine.RecommendationId);
            ProductsWithRecommendations.TryAdd(productIntId, new Dictionary<float, List<int>>());
            ProductsWithRecommendations[productIntId].TryAdd(inputLine.Rank, new List<int>());
            ProductsWithRecommendations[productIntId][inputLine.Rank].Add(recommendationIntId);
        }

        private int InitProduct(string productsId)
        {
            if (productsStr.ContainsKey(productsId))
                return productsStr[productsId];

            productsStr[productsId] = products.Count;
            products.Add(productsId);
            return products.Count - 1;
        }
    }
}